### PR TITLE
Add node metrics in node repo

### PR DESCRIPTION
**Pull request by hakonhall**

### CODE REVIEW

**Review by ldalves — CHANGES_REQUESTED** [review]
I suggest other names for some metrics (in came case, since all other metrics names in the code are in camel case), but overall I don't like camel case in metric names, so would need to change all metrics.  I guess we try to use `.`, but I see we also have `_` in some metrics, if two words are of the same concept.

So, `hostedVespa.activeHosts` change to: `hosted.vespa.active.hosts` or `hosted.vespa.active_hosts` or `hosted_vespa.active_hosts`?  

Better if we use `vespa.hosted.xxx` to match with `vespa.node`.

@yngveaasheim, what do you think? This is related to some of your goals to ensure we have good metrics names as well. :)

@gjoranv could you please run an overall review as well? :)

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/MetricsReporter.java [hunk 1]
```diff
@@ -20,22 +29,138 @@
...
+            context = getContextAt(
+                    "state", node.state().name(),
+                    "hostname", node.hostname());
+        }
+
+        Optional<Version> currentVersion = node.status().vespaVersion();
+        // Node repo checks for !isEmpty(), so let's do that here too.
+        if (currentVersion.isPresent() && !currentVersion.get().isEmpty()) {
+            double currentVersionNumber = getVersionAsNumber(currentVersion.get());
+            metric.set("currentVersion", currentVersionNumber, context);
```
[the remark this answers is not in the document]
Comment (hakonhall) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/MetricsReporter.java]: Done

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/MetricsReporter.java [hunk 2]
```diff
@@ -20,22 +29,138 @@
...
+            metric.set("wantToRestart", wantToRestart ? 1 : 0, context);
+
+            Version wantedVersion = allocation.get().membership().cluster().vespaVersion();
+            double wantedVersionNumber = getVersionAsNumber(wantedVersion);
+            metric.set("wantedVersion", wantedVersionNumber, context);
+
+            Optional<Version> currentVersion = node.status().vespaVersion();
+            boolean converged = currentVersion.isPresent() &&
+                    currentVersion.get().equals(wantedVersion);
+            metric.set("wantToChangeVersion", converged ? 0 : 1, context);
```
[the remark this answers is not in the document]
Comment (hakonhall) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/MetricsReporter.java]: Done

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/MetricsReporter.java [hunk 3]
```diff
@@ -20,22 +29,138 @@
...
+            long wantedRestartGeneration = allocation.get().restartGeneration().wanted();
+            metric.set("wantedRestartGeneration", wantedRestartGeneration, context);
+            long currentRestartGeneration = allocation.get().restartGeneration().wanted();
+            metric.set("currentRestartGeneration", currentRestartGeneration, context);
+            boolean wantToRestart = currentRestartGeneration < wantedRestartGeneration;
+            metric.set("wantToRestart", wantToRestart ? 1 : 0, context);
+
+            Version wantedVersion = allocation.get().membership().cluster().vespaVersion();
+            double wantedVersionNumber = getVersionAsNumber(wantedVersion);
+            metric.set("wantedVersion", wantedVersionNumber, context);
```
[the remark this answers is not in the document]
Comment (hakonhall) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/MetricsReporter.java]: Done

**Comment by hakonhall** [comment]
On formatting of metric names: Either way is fine with me. For these changes I have followed the style of the existing metrics defined by the class.

**Comment by ldalves** [comment]
yes, I noticed that, that's why I left as camelCase in the suggestion as well... changing metrics name for all metrics will require more work..

**Review by ldalves — APPROVED** [review]

**Comment by hakonhall** [comment]
I'll fix any further concerns @yngveaasheim and @gjoranv may have in a follow up PR to get this out.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/MetricsReporter.java [hunk 4]
```diff
@@ -20,22 +29,138 @@
...
     @Override
     public void maintain() {
+        List<Node> nodes = nodeRepository().getNodes();
+        nodes.forEach(this::updateNodeMetrics);
+        updateStateMetrics(nodes);
+        updateDockerMetrics(nodes);
+    }
+
+    private void updateNodeMetrics(Node node) {
+        // Dimensions automatically added: host, vespaVersion, zone, role, and colo.
```
Comment (gjoranv) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/MetricsReporter.java]: What does 'automatically' mean here? If this is done elsewhere, this comment will become outdated when the other code changes.

**Review by gjoranv — COMMENTED** [review]
I assume these metrics are only visible internally. All metrics for external use should be subject to arch review.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/MetricsReporter.java (hunk 4 above)
Comment (hakonhall) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/MetricsReporter.java]: Let me remove the comment.

**Comment by hakonhall** [comment]
> I assume these metrics are only visible internally. All metrics for external use should be 
> subject to arch review.

These metrics will be used to set up an activity graph that may or may not be useful, and will be evaluated at sprint closure.  If useful, we'll make it available in the hosted application dashboard.

What's your definition of "external use"? Any metrics COULD be externally used. For now, these metrics have been requested by and will be used by us, and it will most likely not be useful for normal users. I can bring up this topic on sprint closure unless you clearly see it falls into one or the other category.

**Pull request merged**